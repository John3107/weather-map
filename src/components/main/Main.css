.App {
    display: flex;
    flex-direction: column;
    position: relative;
}

.input {
    white-space: nowrap;
    position: relative;
}

.textField{
    width: 100%;
}

.clear-icon {
    position: absolute;
    color: blue;
    cursor: pointer;
    top: 16px;
    right: 10px;
}

.context {
    width: 100%;
    position: absolute;
    left: 0;
    top: 56px;
}

.city {
    display: flex;
    justify-content: space-between;
    border: 1px solid #c7c3c3;
    padding: 5px;
    font-family: sans-serif;
    font-size: 14px;
}

.cityRightSide {
    display: flex;
    text-align: right;
}

.cityRightSide :first-of-type {
    margin-right: 5px;
}

.city:hover {
    background: #303f9f;
    color: #fff;
    cursor: pointer;
}

:last-of-type {
    border-bottom-left-radius: 5px;
    border-bottom-right-radius: 5px;
}
