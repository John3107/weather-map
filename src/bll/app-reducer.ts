import {CityType, CityWeatherType} from "../types/types";
import {AppThunk} from "./store";
import {weatherAPI} from "../api/api";

const initialState: InitialStateType = {
    cities: [],
    citiesContext: []
}

export const appReducer = (state: InitialStateType = initialState, action: ActionsType): InitialStateType => {
    switch (action.type) {
        case 'CITIES':
            return {...state, cities: action.citiesList}
        case 'CITIES-CONTEXT':
            return {...state, citiesContext: action.citiesContextArray}
        default:
            return state
    }
}

export const initialCitiesDataTC = (): AppThunk => (dispatch) => {
    const getCities = localStorage.getItem('cities-list')
    if (getCities) {
        dispatch(setCitiesAC(JSON.parse(getCities)))
    }
}

export const addCityTC = (newCity: CityType): AppThunk => (dispatch) => {
    dispatch(setCitiesContextAC([]))
    const citiesList = localStorage.getItem('cities-list')
    const citiesListParsed = citiesList && JSON.parse(citiesList)
    weatherAPI.getCityWeather(newCity.lat, newCity.lon)
        .then((res) => {
            const cityWeather = {
                id: String(res.data.lon) + String(res.data.lat),
                city: newCity.name,
                country: newCity.country,
                temp: Math.round(res.data.current.temp / 10.59),
                description: res.data.current.weather[0].description,
                icon: res.data.current.weather[0].icon,
                lat: res.data.lat,
                lon: res.data.lon
            }
            if (citiesListParsed[0]) {
                const searchCity = citiesListParsed.find((el: CityWeatherType) => el.id === cityWeather.id)
                let addCity
                if (searchCity) {
                    addCity = citiesListParsed.map((el: CityWeatherType) => el.id === cityWeather.id
                        ? ({...cityWeather})
                        : el)
                } else {
                    addCity = [...citiesListParsed, cityWeather]
                }
                localStorage.setItem('cities-list', JSON.stringify(addCity))
                dispatch(setCitiesAC(addCity))
            } else {
                localStorage.setItem('cities-list', JSON.stringify([cityWeather]))
                dispatch(setCitiesAC([cityWeather]))
            }
        })
        .catch((err) => {
            console.log(err)
        })
}

export const citiesContextTC = (title: string): AppThunk => (dispatch) => {
    if (!title) return dispatch(setCitiesContextAC([]))
    weatherAPI.getCitiesWeather(title)
        .then((res) => {
            dispatch(setCitiesContextAC(res.data))
        })
        .catch((err) => {
            console.log(err)
        })
}

export const deleteCardTC = (cardId: string): AppThunk => (dispatch) => {
    const citiesList = localStorage.getItem('cities-list')
    const citiesListParsed = citiesList && JSON.parse(citiesList)
    const citiesListFiltered = citiesListParsed.filter((el: CityWeatherType) => el.id !== cardId)
    localStorage.setItem('cities-list', JSON.stringify(citiesListFiltered))
    dispatch(setCitiesAC(citiesListFiltered))
}

export const setCitiesAC = (citiesList: CityWeatherType[]) => ({type: 'CITIES', citiesList} as const)
export const setCitiesContextAC = (citiesContextArray: CityType[]) =>
    ({type: 'CITIES-CONTEXT', citiesContextArray} as const)

export type CitiesActionType = ReturnType<typeof setCitiesAC>
export type CitiesContextActionType = ReturnType<typeof setCitiesContextAC>

export type InitialStateType = {
    cities: CityWeatherType[],
    citiesContext: CityType[]
}

type ActionsType = CitiesActionType | CitiesContextActionType