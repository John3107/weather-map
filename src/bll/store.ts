import {AnyAction, applyMiddleware, combineReducers, legacy_createStore as createStore} from 'redux'
import thunkMiddleware, {ThunkAction, ThunkDispatch} from 'redux-thunk'
import {appReducer} from "./app-reducer";

export const rootReducer = combineReducers({
    app: appReducer
})
export const store = createStore(rootReducer, applyMiddleware(thunkMiddleware));

export type AppRootStateType = ReturnType<typeof rootReducer>

export type AppDispatch = ThunkDispatch<AppRootStateType, unknown, AnyAction>

export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppRootStateType, unknown, AnyAction>