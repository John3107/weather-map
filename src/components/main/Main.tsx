import React, {useEffect, useState} from 'react';
import './Main.css';
import {TextField} from "@material-ui/core";
import Card from "./card/Card";
import {useAppDispatch, useAppSelector} from "../../hooks/hooks";
import {addCityTC, citiesContextTC} from "../../bll/app-reducer";
import {CityType} from "../../types/types";
import {Clear} from '@material-ui/icons';

const Main = () => {

    const data = useAppSelector()
    const dispatch = useAppDispatch()
    const [enteredTitle, setEnteredTitle] = useState<string>('')

    useEffect(() => {
        dispatch(citiesContextTC(enteredTitle))
    }, [enteredTitle, dispatch])

    const selectCityHandler = (selectedCity: CityType) => {
        setEnteredTitle('')
        dispatch(addCityTC(selectedCity))
    }

    const onBlurHandler = () => {
        setTimeout(() => dispatch(citiesContextTC('')), 100)
    }

    return (
        <div>
            <div className="add">
                <div className="input">
                    <TextField className="textField"
                               label="Enter city name"
                               variant="outlined"
                               onChange={(enteredTitle) => setEnteredTitle(enteredTitle.target.value)}
                               onBlur={onBlurHandler}
                               value={enteredTitle}
                    />
                    <Clear className="clear-icon" onClick={() => setEnteredTitle('')}/>
                </div>
                <div className="context">
                    {
                        data.citiesContext.map((el, i) => {
                            return (
                                <div className="city" key={i} onClick={() => selectCityHandler(el)}
                                     id="context-menu">
                                    <span>{el.name}</span>
                                    <div className="cityRightSide">
                                        <div>{el.state}</div>
                                        <div>{el.country}</div>
                                    </div>
                                </div>
                            )
                        })}
                </div>
            </div>
            <div className="Cards">{
                data.cities.map((el) => <Card {...el} key={el.id}/>)
            }</div>
        </div>
    );
}

export default Main;