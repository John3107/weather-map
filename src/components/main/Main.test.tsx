import { Provider } from "react-redux";
// @ts-ignore
import renderer from "react-test-renderer";
import Main from './Main';
import {store} from "../../bll/store";

it('renders correctly snapshot', () => {
    const tree = renderer
        .create(<Provider store={store}><Main/></Provider>)
        .toJSON();
    expect(tree).toMatchSnapshot();
});

