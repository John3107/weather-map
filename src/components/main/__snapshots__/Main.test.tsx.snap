// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="add"
  >
    <div
      className="input"
    >
      <div
        className="MuiFormControl-root MuiTextField-root textField"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink={false}
        >
          Enter city name
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-3"
            >
              <span>
                Enter city name
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root clear-icon"
        focusable="false"
        onClick={[Function]}
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </div>
    <div
      className="context"
    />
  </div>
  <div
    className="Cards"
  />
</div>
`;

exports[`renders correctly snapshot 1`] = `
<div>
  <div
    className="add"
  >
    <div
      className="input"
    >
      <div
        className="MuiFormControl-root MuiTextField-root textField"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink={false}
        >
          Enter city name
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-3"
            >
              <span>
                Enter city name
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root clear-icon"
        focusable="false"
        onClick={[Function]}
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </div>
    <div
      className="context"
    />
  </div>
  <div
    className="Cards"
  />
</div>
`;
